File: stack/broker/audit.js
// Shared audit-log writer for provider files. Baked into the image (unlike
// providers/, which is bind-mounted) so every provider gets it via a plain
// relative require, e.g. `require("../audit")` from /app/providers/foo.js.
//
// Emits one JSON line per event to AUDIT_LOG — never a credential value, only
// the shape of what happened (which provider, which route, cache hit/miss).
// A no-op when AUDIT_LOG is unset, so deployments that don't mount the
// audit-logs volume (e.g. the stack/compose.yaml reference skeleton, unless
// wired up) keep working unchanged.
const fs = require("fs");

const AUDIT_LOG = process.env.AUDIT_LOG;

function logEvent(event, fields = {}) {
  if (!AUDIT_LOG) return;
  const line = JSON.stringify({
    // +00:00 and no milliseconds — matches cred-gateway's nginx
    // $time_iso8601 and the proxy's audit.py exactly, rather than
    // toISOString()'s "...ssZ".
    ts: new Date().toISOString().replace(/\.\d+Z$/, "+00:00"),
    service: "broker",
    event,
    ...fields,
  });
  fs.appendFile(AUDIT_LOG, line + "\n", (err) => {
    if (err) console.error("[broker] audit log write failed:", err.message);
  });
}

module.exports = { logEvent };

File: stack/proxy/audit.py
"""Shared audit-log writer for proxy addons.

Baked into the image at /opt/agent-proxy and put on PYTHONPATH (see
Dockerfile), so any addon in the bind-mounted /addons directory can
`import audit` regardless of which addon file it lives next to.

Emits one JSON line per event to AUDIT_LOG — never a credential value, only
the shape of what happened (which host, which provider, blocked or injected).
A no-op when AUDIT_LOG is unset, so addons that call this keep working in
deployments that haven't wired up the audit-logs volume.
"""
import json
import os
import time

_AUDIT_LOG = os.environ.get("AUDIT_LOG")


def log_event(event: str, **fields) -> None:
    if not _AUDIT_LOG:
        return
    line = json.dumps({
        # +00:00, not a "Z" suffix — matches cred-gateway's nginx
        # $time_iso8601, which has no "Z"-suffixed form.
        "ts": time.strftime("%Y-%m-%dT%H:%M:%S+00:00", time.gmtime()),
        "service": "proxy",
        "event": event,
        **fields,
    })
    try:
        with open(_AUDIT_LOG, "a") as f:
            f.write(line + "\n")
    except OSError as e:
        print(f"[proxy] audit log write failed: {e}")
